raft.go

```diff
@@ -1232,23 +1238,22 @@ func (r *Raft) quorumSize() int {
 	return voters/2 + 1
 }
 
-// appendConfigurationEntry changes the configuration and adds a new
-// configuration entry to the log
-func (r *Raft) appendConfigurationEntry(future *configurationChangeFuture) {
-	configuration := cloneConfiguration(r.configurations.latest) // same as committed
-
-	if future.prevIndex > 0 && future.prevIndex != r.configurations.latestIndex {
+// nextConfiguration generates a new Configuration from the current one and a
+// configuration change request. It's split from appendConfigurationEntry so
+// that it can be unit tested easily. This function reports errors directly to
+// the future; it returns an empty configuration if and only if
+// future.responded is true.
+func nextConfiguration(current Configuration, currentIndex uint64, future *configurationChangeFuture) Configuration {
```

Comment:
Yeah agree the future at this level is a bit ugly. I'd be in favor of a nested struct in the future as the argument to this with `(Configuration, error)` as the return so it's more natural to call this.